0017-ilp-errors/0017-ilp-errors.md

```diff
@@ -0,0 +1,73 @@
+# Interledger Protocol Errors
+
+This spec defines a standard error format and codes to be used when ILP payments are rejected.
+
+Inspired by [HTTP Status Codes](https://tools.ietf.org/html/rfc2616#section-10), ILP errors are categorized based on the intended behavior of the caller when they get the given error.
+
+## Error Propagation
+
+ILP connectors MUST relay errors when they are notified of a transfer being rejected.
+
+Connectors SHOULD NOT modify the error codes or messages except in the following way: connectors MAY append their ILP addresses, separated from the error message by a newline (`\n`), to assist with debugging/tracing.
+
+## Data Format
+
+ILP errors are UTF-8 encoded strings that are a maximum of 65535 bytes long. Error messages that are too long for a ledger to handle SHOULD be truncated to the first 65535 bytes.
+
+Errors consist of a three character code, followed by a colon (`:`) and an arbitrary message set by the entity that raised the error. Transport and application layer protocols MAY define error messages that have special significance in their context.
```

Comment:
What happens for asynchronous error events and/or non-HTTP protocols?. I'm thinking of a wallet-client with just a websocket connection to the ILP-ledger. Such wallet-client sends an HTTP REST request with an "accept payment". The initial HTTP request is validated and forwarded (no-error is returned). The final recipient notice an S?? error later on (wrong condition for example) and wants to propagate it to the originating wallet. For the client to be notified all the information must be encapsulated in an (JSON?) object since the websocket channel has no notion of request headers.

Mabye will more conservative to define first the information the error message (serializable object in OOP terminology) must carry and define then, for each type of transport (syncrhonous HTTP request/response, websockets, gRPC, ...), the codification used to transmit it (header + json, json alone, any other future format, ...)? (or maybe I'm completly lost)